src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/ProjectDependencyHelper.cs

```diff
@@ -0,0 +1,145 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using System.Runtime.Serialization;
+using Microsoft.CodeAnalysis.LanguageServer.LanguageServer;
+using Microsoft.CodeAnalysis.MSBuild;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.Extensions.Logging;
+using NuGet.ProjectModel;
+using NuGet.Versioning;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.HostWorkspace;
+internal static class ProjectDependencyHelper
+{
+    private const string UnresolvedProjectDependenciesName = "workspace/_roslyn_projectHasUnresolvedDependencies";
+
+    internal static bool HasUnresolvedDependencies(ProjectFileInfo newProjectFileInfo, ProjectFileInfo? previousProjectFileInfo, ILogger logger)
+    {
+        if (previousProjectFileInfo is null)
+        {
+            // This means we're likely opening the project for the first time.
+            // We need to check the assets on disk to see if we need to restore.
+            return CheckProjectAssetsForUnresolvedDependencies(newProjectFileInfo, logger);
+        }
+
+        if (newProjectFileInfo.TargetFramework != previousProjectFileInfo.TargetFramework)
+        {
+            // If the target framework has changed then we need to run a restore.
+            return true;
+        }
+
+        var newPackageReferences = newProjectFileInfo.PackageReferences;
+        var previousPackageReferences = previousProjectFileInfo.PackageReferences;
+
+        if (newPackageReferences.Length != previousPackageReferences.Length)
+        {
+            // If the number of package references has changed then we need to run a restore.
+            // We need to run a restore even in the removal case to ensure the items get removed from the compilation.
+            return true;
+        }
+
+        if (!newPackageReferences.SetEquals(previousPackageReferences))
```

Comment:
We should really just teach SetEquals to do the length check first, but that can be later.